0726-dom-element-descriptor-interface.md

```diff
@@ -0,0 +1,265 @@
+
+## Motivation
+
+In this RFC we'll use **DOM element descriptor** to mean a Javascript value that can be used to retrieve one or more `Element`s from a rendered DOM. Examples include a string containing a CSS selector, a direct reference to an `Element`, a function that returns an `Element` or `null`, a function that returns an array of `Element`s, etc.
+
+In this RFC we'll use **DOM helper** to mean an API that is passed a DOM element descriptor, and performs some actions based on the element or elements the descriptor describes. Examples include `@ember/test-helpers`' DOM helper methods such as `click()` and `triggerEvent`, and `qunit-dom`'s assertion factory API, i.e. `assert.dom()`.
+
+Ember's testing ecosystem has converged on a single set of recommended DOM helpers. `@ember/test-helpers` provides the core value of implementing DOM interactions, while `qunit-dom` provides the core value of making assertions about DOM query results. They do _not_ (nor should they) provide the core value of advanced DOM querying -- they act on DOM elements, and do the simple baseline of supporting two kinds of DOM element descriptors, CSS selector strings and `Element` references.
+
+[Page Objects](https://martinfowler.com/bliki/PageObject.html) are a powerful and popular tool for making test code well organized, readable, and maintainable. The core value they provide is supporting the implementation of a domain-specific languange (DSL) for interacting with a rendered DOM. Such DSLs can be thought of as, among other things, providing powerful and resuable DOM element descriptors.
+
+Supporting a clean integration of page objects (or, indeed, anything that produces DOM element descriptors that aren't CSS selectors or direct `Element` references) with DOM helpers like the ones in `@ember/test-helpers` and `qunit-dom` would keep each focused on its core value, while providing "first-class" support for page objects and other DOM element descriptors.
+
+As a case study (and most proximate motivation for this RFC), let's consider `fractal-page-object`. Currently `fractal-page-object` integrates with `@ember/test-helpers` and `qunit-dom` by exposing an `element` property on all page objects, and that works because their APIs accept direct `Element` references. But this has three major drawbacks:
+
+1. It only supports single elements. While this isn't an issue with the `@ember/test-helpers` DOM APIs, in `qunit-dom` that means there's no way to use the multi-element assertions, analogous to  `assert.dom(document.querySelectorAll('.list-item')).exists({ count: 3 })`.
```

Comment:
I've just checked the "qunit-dom" API, and seems there are only 2 methods that support many items, if I'm not missing something, these are `exists(` and `isVisible(`.

This alignes well with my  personal testing experience. Most of the time you are interested to interact or assert against exactly 1 element. If test methods forgives you an unintended multiple selection, but do only select the first element for assertion, it could lead to some mistakes in tests. This is the reason, why in the ember-cli-page-object the "find-exactly-one-element-or-fail" approach is leveraged actively.

That being said, I think exposing `elements` at this point, is not necessarily a good thing to do. Maybe we could reduce the scope of the RFC and implementation for now, by sacrificing the `elements`, and focus on a minimal but the most common flow for a single `element` with the `description`?

Reply:
A couple of thoughts:

1. While there are only 2 methods that support multi-element matching, in my experience it's a pretty common usage. I personally do some variant of `assert.dom('.list-item').exists({ count: 3 })` quite often. When using page object libraries it may be less common because you would just do `assert.equal(page.listItems.length, 3)`, but that results in less informative assertion messaging (because `assert` only knows it was given two `Number`s, and nothing about the selector used to match the elements). I think that's not a big problem, but one of the goals of this RFC is to enable further convergence on `qunit-dom`, with all its well thought out semantics and messaging, even when using page objects.
2. I'm reluctant to scope this down to only single-element scenarios, because then the integration with `qunit-dom` gets a lot more difficult to think about. At the point that you call `assert.dom(...)` we don't know if you're planning to do a single-element access, a multi-element access, or both! For example, `assert.dom('.thing').exists({ count: 1 }).hasClass('foo')`. So if we were passing objects to `assert.dom()` that could only provide single elements, I'm just not sure how we'd make that work. I guess `qunit-dom` would have to throw an error at the time that the user calls `.exists({ count })`, and that seems pretty far from ideal.

I am very interested in making sure this proposal would support your "find-exactly-one-element-or-fail" and I think it does? If I understand it, you do that enforcement when you are doing the element lookup, right? So I believe you could implement your page objects to produce element data whose `element` getter verifies that the page object's selector only matches one element, and whose `elements` getter unconditionally throws an error (assuming `multiple: true` wasn't set).

Reply:
IMO the multiple elements topic can be tricky sometimes. It might even have some extra design. We've deprecated the `multiple` in the ember-cli-page-object just because it overcomplicated internals, w/o any value _(cause everything is covered with collections)_. Unfortunately, I don't have a single good answer for you, so let me drop some raw thoughts here.

Let's consider your example modified to have `count: 3`:

```
 `assert.dom('.thing').exists({ count: 3 }).hasClass('foo')`
```

I'd expect it to evaluate `hasClass` over each of 3 elements(is it just me? 😄). But documentation says:

> assertion will return true if **any** of the element's CSS classes match.

However, _(here, I have to admit I'm not an active qunit-dom user at the moment so I may miss smth)_ if I read the `hasClass` [implementation](https://github.com/simplabs/qunit-dom/blob/master/lib/assertions.ts#L627) correct, it just checks only for the first element, and this is unexpected to me. Here I just want to highlight a non-obvious hidden complexity of the MANY thing, so even in this small use case, I am not sure, as a user, what to expect when it comes to a `mutliple` thing.

The other question I have is, whether each `elements` item needs it's own `description`. I'm not currently sure, how would I use it, but I can imagine, I would want to specify an order for each element in the list, so the final error message would have an exact index per each failed element(let's imagine for a moment `hasClass` iterates over each element).

Also it just feels a little bit undermenistic to me to put both single `element` and many `elements` in the same descriptor. For instance, in terms of the ember-cli-page-object, I think I'd prefer to return an `element` for a node and `elements` for collection. I mean, we may want to distinct between the single VS multiple descriptiors. This one is not a big deal to me, but the need for a mixed approach is not clear to me.

> I am very interested in making sure this proposal would support your "find-exactly-one-element-or-fail" and I think it does? If I understand it, you do that enforcement when you are doing the element lookup, right? 

Yes, you are right. This RFC doesn't block this behaviour in any way. I've mentioned it to tell that the `element` semantics is much more obvious, and it's simpler to reason about, at least for me.

----

That being said, I'm not strongly opposed to the `elements` in general. I'm just uncertain if the current design is good enough. And unfortunatelly for now, we have too few use cases for it, so it seems tricky to reason about the real requirements. Maybe we can invent a syntetic test helper for demonstrational purpose(RFC only), which will do something more than just `exists`. Something like `hasClass` iterating over all the elements, with error/success message? I think it may help us to design the thing with more considerations in mind.

Reply:
Yeah, I definitely agree with you that multi-element semantics are kinda complex and often not completely clear, and your `count: 3` example is a good illustration of that.

The documentation is a little confusing, but I think accurate as long as you're a bit of a grammatical pedant 😆  `any of the element's CSS classes` refers to any of the CSS classes of the (singular) element. `any of the elements' CSS classes` would refer to any of the CSS classes of the (plural) elements. So the documentation already assumes that there's only one element we're talking about, which I think of/refer to as a `single element context`.

But yes, this is a good illustration of the ambiguity and confusing-ness of single vs. multiple elements. But that has its root in a very broad ecosystem-wide pattern, which is CSS selectors. We use selectors all over the place to reference the DOM, but they don't encode any information about whether they reference a single element or multiple elements -- it depends on whether you pass them to `querySelector()` or `querySelectorAll()`.

So any DOM helper that accepts a selector string as an argument embraces this ambiguity, and either solves it with implicit behavior (`@ember/test-helpers` and most of `qunit-dom` implicitly look for the first match), explicit behavior (those two `qunit-dom` methods accept an argument that tells them if they are in single-element mode or multi-argument mode), or via enforcement (e.g. `ember-cli-page-object` interprets them as single-element references and enforces the validity of that with an assertion).

This RFC is not seeking to address that, because that would involve breaking changes to `qunit-dom`, but also because I think it's fundamentally un-addressable because this RFC is seeking to "extend" CSS selectors, and CSS selectors are fundamentally ambiguous as far as referencing one element or multiple elements.

I think this also hides the ambiguity from most end users, because they will never see the `element` and `elements` properties. They'll probably use page objects that automatically integrate with DOM helpers, and that they will use in the same way they use CSS selectors today, or else they'll use the factory helpers described in this RFC and will probably 95% of the time pass in a single element to the factory function without needing to know about multi-element support.

Reply:
So that's a bunch of reasons, but the bottom line is that I tried to mirror how CSS selectors work and are used, in large part because we have such a limited set of use cases, so I prioritized parity over re-imagining anything. Specifically:

* a descriptor is equivalent to a CSS selector
* accessing `.element` on the descriptor's data is equivalent to calling `document.querySelector()` on a CSS selector
* accessing `.elements` on the descriptor's data is equivalent to calling `document.querySelectorAll()` on a CSS selector
* passing a descriptor to a DOM helper is equivalent to passing a CSS selector to a DOM helper, with exactly the same ambiguities

Maybe that also helps to answer `The other question I have is, whether each elements item needs it's own description`. To make that work, I'd expect something outside of this RFC. For example, in `fractal-page-object`, if you were to do the odd `assert.dom(page.listItems).hasClass('foo')` the description would be something like `.list-item` because that's the selector corresponding to the `page.listItems` node. If you did `assert.dom(page.listItems[2]).hasClass('foo')` the description would be something like `.list-item[2]` because that's the a decent selector-ish representation of the list item and index 2. So I'm not currently thinking of multi-element scenarios as operating on the contents of the list of Elements, but as operating on the list of Elements itself (e.g. `.exists({ count })` is asking how long the list is, not anything about the elements in the list).

So, rather than using a single descriptor to reference multiple elements and give each of them their own description, I'd expect a consumer to construct a descriptor, with description, for each element -- the difference between "talking about" a list and talking about the contents of the list.

Reply:
That's my thinking here. I do want to pause and give you a big thanks for putting time and thought into this! I'm still feeling pretty confident that descriptors need to embrace the single/multi element ambiguity so they are fundamentally at parity with CSS selectors, but I'm only medium confident that we've got the right design for that in this RFC, so appreciate challenges/ideas/etc.

Reply:
Let's say we have a page object collection with the `ul > li` selector for items. 

Now coming back to the hypothetical `hasClass` example, when passing the whole collection:

```js
  assert.dom(MyCoolList.items).exist({ count: 2 }).hasClass('some-class');
```

if the second item fails to satisfy the condition, It'd be nice to see as pricese context as possible in the error message, like:

>  does not have a class "some-class" for:
> 
> Page Object: MyCoolList.items[2]
> Selector "ul > li[2]" 

_(ya! this is a bit overwhelmed with context 😅 )_

I think if we don't have a `description` per `element`, we can't have such a level of control in the error messages, and we can only have, smth like:

> does not have a class "some-class" for "ul > li" 

Through there is not such a thing like multiple `hasClass` yet, I hope my example makes sense for you.

Reply:
Yeah, that makes sense. Although in that case it would be `qunit-dom` that would be implementing the iteration over the elements, so I would think that it would be its job to generate a useful message based on the description of the whole collection plus its knowledge of the indices of the elements, e.g.

`class 'some-class' not present on 'ul > li' collection at index 2`

where it gets `ul > li` from the descriptor and constructs the rest of the message itself. That way it could also do

`class 'some-class' not present on 'ul > li' collection at index 2, 4, and 5`

whereas if the descriptor provides a description for each element including the index, all `qunit-dom` could do is

`class 'some-class' not present on 'ul > li[2]', 'ul > li[4]', and 'ul > li[5]'`

which could quickly get very hard to read if the selector was something like `.page .content-section .accordion-frame[2] ul > li` 😅 . If one description is provided for the whole collection, then since the DOM helper that's using the collection knows the index of each item as it iterates, it could construct either of the messages above, whereas if we make descriptors provide a per-element description, then the DOM helpers have less flexibility in their messaging.

We could certainly do both -- have descriptors provide a description for the collection, and optionally provide a parallel array of descriptions or a function that is given an element and returns a description. But this could also be added later once we have gather more usage data to determine if it would be helpful. I think either way we'd want a whole-collection description so `exists({ count })` can output `expected 2 elements in 'ul > li' collection but found 3` or whatever, so I guess I'm proposing that we defer the per-element description question until we have more data indicating a need for it.

Reply:
Yes, I agree about the qunit-dom responsibility here. Thanks for pointing me to it!